### DISCUSSION TITLE
Specifying commands on using a new path for terminal tool

### DISCUSSION THREAD

**Discussion by goyalyashpal** [geany/geany#3303] [General]
### TL;DR
* any known working alternate to the path for terminal: `cmd.exe /Q /C %c` ?
* ... on geany 1.38 on windows 10 20H2
* bash via msys2 is available too

+ 'was thinking about mintty (via msys2) but obviously, other terminals are welcome too
+ terminals with easy theming support are preferred

- tangentially: similar to `libexec/geany/geany-run-helper.bat`; what is the run-helper file for linux/unix-like systems?

----
* search term: changing terminal for run/execute
* For -vv (verbose verbose 😆) version of this, see the following ~~hidden~~ comment

**Comment by goyalyashpal** [comment]
~~_will hide this comment myself as duplicate_~~

----

[[_TOC_]]

### Summary
* Can't figure out what commands to specify on using a new path for terminal tool
* Was trying to switch to some terminal which satisfies my requirements:

### Questions:

* Is the command for making it run on some other terminal known?
* What is the helper file on linux? thinking about: maybe it's easier to call bash on mintty than cmd on mintty

### Details
#### Thanks
* I quite like the "F5 to run" option in geany, \
how it abstracts the running procedure for stuff with that `geany-run-helper.bat` file[^helperfile], it's nice
* and even more so the super awesome "Build Menu"[^build] including predefined rules, as well as "make with custom targets". thanks a lot for such efficient and flexible implementation 😊

[^build]: `C:/Program Files/Geany/` `share/doc/geany/html/index.html#build-menu`
[^helperfile]: `C:\Program Files\Geany\` `libexec\geany\geany-run-helper.bat`

#### Problem
* for "f5 to run", as the command `cmd.exe /Q /C %c` [^termpath][^cmdflags] shows already, it runs with `cmd` <-- as also suggested by extension `.bat` i guess
* and `cmd` runs by default with `conhost` , which is not easily themeable. That's why I want to change it.

[^termpath]: Edit : Preferences (alt ^p) > tools > tools path : terminal
[^cmdflags]: `/Q` echo off; `\C` carryout the string as command and terminate <-- `cmd /?` <-- [edu4java] 



[edu4java]: http://edu4java.com/en/concepts/ms-dos-commands-modifiers-parameters.html

#### Tries & Attempts
* I tried theming conhost with colortool[^ct] but, suffice to say, it didn't work
* To change to mintty (installed via Msys), I thought directly putting `mintty -e`  as in: `mintty -e cmd.exe /Q /C %c` or `mintty -e cmd.exe /Q /C "%c"` but it didn't work
* thought about using the nested (and hence, escaped) quotes[^escquote] - don't remember how it went

[^escquote]: https://github.com/geany/geany/issues/1757#issuecomment-619448490
[^ct]: [src/tools/ColorTool] <-- [README.md#11] <-- [MS devblogs] <-- ddgo search : [light theme for conhost]

[src/tools/ColorTool]: https://github.com/microsoft/terminal/tree/main/src/tools/ColorTool
[MS devblogs]: https://devblogs.microsoft.com/commandline/introducing-the-windows-console-colortool/
[light theme for conhost]: https://duckduckgo.com/?q=light+theme+for+conhost&ia=web
[README.md#11]: https://github.com/microsoft/terminal/blob/e1e6e662f9d00d55770c2c947066d7de18ea286f/README.md?plain=1#L11

#### Teminal's requirements

| requirements | mintty | conhost | others? 
| --- | --- | --- | ----
| customisable font | v | v |
| the font size corresponds one-to-one | v | **XX** |
| default window size can be easily specified | v | v |
| easily themeable | v | **XX** |

### Environment
* OS: Windows 10 20H2
* Geany: 1.38 "Sulamar" (built on or after Oct  9 2021)
* Mintty: mintty 3.6.1 (x86_64-pc-msys)

**Comment by elextr** [comment]
The `preferences->tools->terminal` command is the one used to run programs (`build->execute`) in a terminal emulator application (ie one that shows a window on the screen) so any messages are visible. As its a terminal `bash` is not applicable, for eg the default on Linux is `xterm`.  

The script `geany-run-helper.bat` is what is passed as the command (ie the %c in `tools->terminal`) with arguments specifying the directory to run in, if the terminal should autoclose when the program exits or keep the terminal open so the user can read error messages, and the command to run.  So any replacement for `cmd.exe` must interpret `.bat` scripts or `geany-run-helper.bat` needs to be re-written to an alternate interpreter.  This is left as an exercise for the reader.

For hysterical (sorry that should read historical) reasons on Linux a new temporary "sh" run script is written every execute so there is no Linux equivalent of `geany-run-helper.bat`.

Macos is a whole 'nother world, we won't talk about it.

**Comment by goyalyashpal** [comment]
> As [the requirement?] its a terminal `bash` is not applicable

* if you are talking about the requirement being of a terminal like xterm, 
* then, `cmd` is not a terminal either. it's a kind of shell in windows. 
* the terminal on windows in `conhost`. 
* you can verify this by running both these commands in other terminals, `cmd` will start this shell in that terminal, while `conhost` will launch a new window for this terminal.

> a new temporary "sh" run script is written every execute

can u share that itself?

> This is left as an exercise for the reader.

what? the "re-writing of the script" or "the replacement for the whole command: `cmd.exe /Q /C %c` "?

**Comment by elextr** [comment]
>  cmd is not a terminal either.

IIUC `cmd` uses Windows Console to display output and get user input from a window if it is not already in one, so it is in fact a combination of terminal and shell.  IIUC Geany is a GUI program, so it doesn't have a console, so when it spawns `cmd` a new console is opened.  Launching `cmd` in an existing console obviously uses that console.

> > a new temporary "sh" run script is written every execute

> can u share that itself?

https://github.com/geany/geany/blob/b9891f2912a7e7374e3121f74b587af1e7b3901d/src/build.c#L1161

> what? the "re-writing of the script" or "the replacement for the whole command: cmd.exe /Q /C %c "?

all of the above

**Comment by goyalyashpal** [comment]
> it is in fact a combination of terminal and shell

I think it's more like a combo of a shell, and a starter script pointed at the terminal `conhost`, but yeah, i get what u want to say

> all of the above

well 😅 , the whole point of this issue was asking if somebody knows the replacement for the command. i know that users have to do this on their own, but i was asking it as a community help sort of thing. 

on a hindsight, yes, this is better suited as a "discussion" - sorry to file it as an issue. can u move it to discussions?

**Comment by elextr** [comment]
> well , the whole point of this issue was asking if somebody knows the replacement for the command. i know that users have to do this on their own, but i was asking it as a community help sort of thing.

Ok, but thats not really clear, maybe restate _just_ the question simply and clearly, "Has anybody or does anybody know how to ...."

> on a hindsight, yes, this is better suited as a "discussion" - sorry to file it as an issue. can u move it to discussions?

done

**Comment by elextr** [comment]
Note that on Linux the shell is specified by the shebang in the helper script, so Geany has no setting to allow it to be specified.

> maybe restate just the question simply and clearly

And maybe describe your use-case, what (not how) you want to do and why.

**Reply by goyalyashpal** [reply]
> And maybe describe your use-case, what (not how) you want to do and why.

yes, i wanted to do that, but got lost in technicalities while filing the issue, that it got mingled in the long version, and got skipped in the TL;DR version 😅 . anyways, following is it:

* i have to take screenshot of the output, and embed those in a document
* for that to be straightforward, the output has to launch in a fixed window width, and in a fixed font size - both of which are compulsory, and have to be easily specifiable
* optionally, it would be nice if the output is in light theme - so that while taking the print out (i.e. hard print out on paper), it looks nice while consuming less ink

These above are requirements, and following is why the current situation is unable to achieve the last optional point of theming:
* output of geany is shown on conhost

<details>
<summary>

* changing theme on conhost is not easy (click for details)
</summary>


* i tried doing that via colortool (see [above comment](https://github.com/geany/geany/discussions/3303#user-content-fnref-ct-f4953bd48a5c2cd01e758990f3d58600)) which required using either of [4-bit terminal](http://ciembor.github.io/4bit/) or [terminal.sexy](https://terminal.sexy/)
* the 4-bit's presets were nice, only required two clicks : white/bright-white in background, and black in foreground
    + screenshot: ![image](https://user-images.githubusercontent.com/19423063/192133055-62147d41-0efe-44b1-84f3-024f97841b09.png)
    + but i just couldn't get the generated .itermcolors (by copy pasting the data:url and decoding it from url-encoding) file to work with conhost. 
    + same with terminal.sexy

</details>

* that is, i can't get a functional light theme for conhost
* and don't know how to change geany to output in a terminal for which it is (easy to do all above)

So, for the time being, 
* i have settled to taking screenshots in the default pure black theme, and doing color inversion on the screenshots. 
* this obv involves one additional postprocessing step, 
* but it's not a ~~big~~ Huge deal once you figure out how to batch-process it with image-magick tool. 
* i still haven't figured out how that will be automated as don't have time, but it's okay for now.

**Reply by goyalyashpal** [reply]
> _on Linux the shell is specified by the shebang in the helper script, so Geany has no setting to allow it to be specified_

ohw, it really skipped from my mind. thanks for reminder.

**Comment by goyalyashpal** [comment]
so, my use-case this time is to make the `execute` command in geany to use a terminal which has [iterm image protocol](https://wezfurlong.org/wezterm/imgcat.html?h=image) support (like wezterm)

currently, geany opens with conhost, and conhost doesn't support that.
